components/logins/src/migrate_sqlcipher_db.rs

```diff
@@ -78,12 +78,60 @@ impl Add for MigrationPhaseMetrics {
     }
 }
 
-pub fn migrate_sqlcipher_db_to_plaintext(
+pub fn migrate_logins(
+    path: impl AsRef<Path>,
+    new_encryption_key: &str,
+    sqlcipher_path: impl AsRef<Path>,
+    sqlcipher_key: &str,
+    salt: Option<String>,
+) -> Result<String> {
+    let path = path.as_ref();
+    let sqlcipher_path = sqlcipher_path.as_ref();
+
+    // Need to make sure we have an old db and new one before we start anything
+    let result = match (path.exists(), sqlcipher_path.exists()) {
+        (true, true) => migrate_sqlcipher_db_to_plaintext(
+            &sqlcipher_path,
+            &path,
+            sqlcipher_key,
+            new_encryption_key,
+            salt.as_ref(),
+        ),
+        (false, true) => match sqlcipher_path.to_str() {
+            None => throw!(ErrorKind::InvalidPath(
+                sqlcipher_path.as_os_str().to_owned(),
+            )),
+            Some(s) => throw!(ErrorKind::InvalidDatabaseFile(s.to_string())),
+        },
+        (true, false) => match path.to_str() {
+            None => throw!(ErrorKind::InvalidPath(
+                sqlcipher_path.as_os_str().to_owned(),
+            )),
+            Some(s) => throw!(ErrorKind::InvalidDatabaseFile(s.to_string())),
+        },
+        (false, false) => throw!(ErrorKind::InvalidDatabaseFile(
+            "Both paths failed".to_string(),
+        )),
+    };
```

Comment:
Do we want to require that the consumer to figure out if a migration should run or not?  It seems simpler for them to always call the function, then we decide if a migration needs to run.  IOW, I think `(_, false)` should be a no-op rather than an error.

I kind of think that we should create the new DB file, rather than requiring the consumer to.  That makes it easier to identify the case where we tried to run a migration and crashed in the middle of it.  In other words:
  - `(true, false)` should be the normal path
  - On `(true, true)`, we should run some code to handle the crashed-in-the-middle case.  Maybe that's deleting the new DB before trying to migrate or maybe it's just deleting both and starting from scratch.

Reply:
🤔 I'm not opposed to this approach but if we decide to handle error cases internally, how will that be communicated to the consumer?

In iOS when a sqlcipher database can't be opened the db file is moved and the storage layer `open` function is called again to create a new db file. I'm not sure this is the worthwhile because of course the data contained in the moved db file is useless, but at least the consumer is aware of what has occurred.

Reply:
Good question.  It seems like this depends on at least a couple things: how are consumers going to handle the error cases and do they need to log/report the result.  Maybe we should figure that out first.  This makes me lean towards merging the current code while we discuss how things ultimately should work.

Reply:
Agreed since there is a bit more nuanced to figuring that part out due to two consumers with potentially different ways of handling. I'll add test cases then we'll get this merged.

Reply:
Sounds good. FWIW on the iOS side sentry logging occurs when the old db file is moved in the scenario I described above so I'm assuming (regardless of whether the old file is deleted or moved) they'd want to do the same if the migration fails.